### ISSUE TITLE
Reusing local variable slot causes decompilation error

### ISSUE THREAD

**Issue by rehwinkel** [leibnitz27/cfr#231]
# CFR version
CFR 0.151

# Compiler
javac 15.0.2

# Description
When compiling this code:
```java
import java.util.ArrayList;

public class Test {
    public static void main(String[] args) {
        {
            ArrayList x = new ArrayList();
        }
        String y = "blub";

        Thread t = new Thread("hi") {
            public void run() {
                System.out.println(y);
            }
        };
    }
}
```

The javac compiler is smart enough to realize that it can reuse local variable slot 1 for *x* and *y*. The resulting decompiled code is this:
```java
import java.util.ArrayList;

public class Test {
    public static void main(String[] stringArray) {
        Object object = new ArrayList();
        object = "blub";
        Thread thread = new Thread("hi", (String)object){
            final /* synthetic */ String val$y;
            {
                this.val$y = string2;
                super(string);
            }

            @Override
            public void run() {
                System.out.println(this.val$y);
            }
        };
    }
}
```

By reusing the local variable index, the decompiler assumes that both of these are referring to the same variable. This causes the anonymous inner class initialization to be screwed up completely.

An idea for fixing this would be to find the lifetime of a given variable. If a local variable index is used twice but overwritten before being used again, there is no reason why the previous use of the variable should be taken into consideration. 

Aditionally to fixing this error, the decompiler could also infer the types of local variables more accurately and doesn't have to use "Object" anymore.

**Comment by leibnitz27** [comment]
TL;DR - Liftetime analysis / reverse single static assignment is a giant PITA :)

Yup - slot re-use is an absolute $£$£"$$£"$£"$".

Cfr actually has a lot of code to cope with this, but consider the following two code snippets (ignore the possibility that we have local variable tables and local type tables, because I'm paranoid, and assume they're full of lies ;) )

int x = 2;
int x2 = x+1;
int x3 = x2+1;
int x4 = x3+1;
int x5 = x4+1;

vs

int x = 2;
x = x+1;
x = x+1;
x = x+1;
x = x+1;

this is of course a trivial example, in which the (provable) type of the slot is unchanged. (actually, it could be a short or a byte, but natch, pedantry).

(this obviously could be improved, but 'discover lifetimes' is more simplistic might be immediately obvious ;) )

**Comment by leibnitz27** [comment]
FWIW - javac is generally *NOT* smart enough to properly handle slot reuse - thankfully - slots are considered seperately within each scope block, NOT each lexical sub-block.  (this means that the two above examples actually generate (or did last time I looked! :) ) different code.

**Comment by x4e** [comment]
@leibnitz27 Local variable slots will only change type when explicitly written to as the new type. Any execution flow after that will use the new type (unless again written to as a new type).

And for reusing local variable slots for the same type, I think it is probably just fine to just use the same name. I cannot really think of cases where this would be incorrect or less readable.

**Comment by Marcono1234** [comment]
Maybe related to #150
